cli/src/diff_util.rs

```diff
@@ -1128,27 +1153,73 @@ pub fn show_diff_summary(
     mut tree_diff: TreeDiffStream,
     path_converter: &RepoPathUiConverter,
 ) -> io::Result<()> {
+    #[derive(Debug)]
+    enum Type {
+        Copy(String, String),
+        Rename(String, String),
+        Modification(String),
+        Addition(String),
+        Deletion(String),
+    }
+
+    let mut file_modifications = Vec::<Type>::new();
+    let mut unresolved_renames = HashMap::<String, usize>::new();
```

Comment:
Before this commit, `jj diff -s -r <huge commit>` would start producing output as soon as possible, but now we need to iterate through the whole diff before we produce any output. The simplest solution is to simply not distinguish between `C` and `R`. I think that's what Mercurial does. If we/you care enough to distinguish them, perhaps we could look ahead in the target tree but that's more expensive and more work, so I would personally leave that for later.

FYI, this is especially unfortunate at Google where it can take a while to produce a large diff of uncached trees. I think it's especially useful to produce output incrementally as a hint that users are doing something they didn't intend when they accidentally produce a large diff (e.g. `jj diff --from cl/12345 --to cl/123456` when they meant `--from cl/123450` or something).

Reply:
`jj diff --stat` already has exactly the behavior you are talking about of walking everything before producing output, and it seems to do fine with it.  We can drop the thing that distinguishes 'C' from 'R' but if feels nice to match git's capabilities here.

Reply:
Yes, `jj diff --stat` does work at Google, it's just that it's silent for a long time if the diff is large. It gets extra slow because it needs to calculate the file diffs too. To be fair, it might be that finding copies between two commits far apart will also be slow, so it might still take a bit of time even if we don't consume the whole tree diff before we produce output.

If you don't mind, I would prefer to see it implemented like this:

1. Find copies
2. For each source path, check in `to_tree` if it's no longer there and then add it to a `deleted_sources` set
3. Walk tree diff and use `deleted_sources` to determine if a copy target is actually a rename

Reply:
Fixed!